# Metagenomics
## Load raw data
```{r loaddata, comment="", message=FALSE, warning=FALSE, eval=FALSE}
counts_raw <- read_tsv("resources/metagenomics/all_species_unfiltered_count_table.txt") %>% 
  rename(genome=1)

genomes_length <- counts_raw[,c(1,4)] %>% 
  rename(length=2)
```

## Read counts
```{r melt, comment="", message=FALSE, warning=FALSE, eval=FALSE}
## Melt dataframe
counts_raw1 <- counts_raw                  # Duplicate data
colnames(counts_raw1) <- gsub("\\.", " ", colnames(counts_raw1))

molten_table <- counts_raw1 %>%
  pivot_longer(!genome,
               names_to = c("sample", ".value"),
               names_sep = " ") %>%
  dplyr::rename("read_count" = "Read") %>%
  dplyr::rename("covered_fraction" = "Covered") %>%
  dplyr::rename("MAG_length" = "Length")

read_counts <- molten_table %>%
  select(genome, sample, read_count) %>%
  pivot_wider(names_from = "sample", values_from = "read_count")%>% 
  column_to_rownames(., "genome") %>%
  filter(rowSums(. != 0, na.rm = TRUE) > 0) %>%
  select_if(~!all(. == 0))%>% 
  rownames_to_column(., "genome")
```

## Genome counts
```{r coverage, comment="", message=FALSE, warning=FALSE, eval=FALSE}
genome_counts_filt <- molten_table %>%
  mutate(read_count = if_else(covered_fraction < 0.3, 0, read_count)) %>% #normalise by read/MAG length
  mutate(read_count = read_count*150/MAG_length) %>%
  select(genome, sample, read_count) %>%
  pivot_wider(names_from = sample, values_from = read_count) %>%
  rename(genome=1)%>% 
  column_to_rownames(., "genome") %>%
  filter(rowSums(. != 0, na.rm = TRUE) > 0) %>%
  select_if(~!all(. == 0)) %>% 
  rownames_to_column(., "genome")

readlength=150
genome_counts <- read_counts %>% 
  arrange(match(genome,genomes_length$genome))%>%
  mutate(across(where(is.numeric), ~ . / (genomes_length$length / readlength) ))%>% 
  column_to_rownames(., "genome") %>%
  filter(rowSums(. != 0, na.rm = TRUE) > 0) %>%
  select_if(~!all(. == 0)) %>% 
  rownames_to_column(., "genome")
```

## Sample metadata
```{r loadmetadata, comment="", message=FALSE, warning=FALSE, eval=FALSE}
sample_metadata <- read_tsv("resources/metagenomics/metadata.tsv") %>%
  rename(sample=1)
genome_counts_filt_r <- genome_counts_filt %>% 
  column_to_rownames(., "genome")
sample_metadata <- sample_metadata[match(colnames(genome_counts_filt_r), sample_metadata$sample), ]
```

## Genome metadata
```{r loadgenomedata, comment="", message=FALSE, warning=FALSE, eval=FALSE}
genome_metadata <- read_tsv("resources/metagenomics/all_together_gtdbtk.tsv") %>%
    rename(genome=1)%>%
  mutate(genome = str_replace(genome, "\\.fa$", ""))%>%#mutate(genome = gsub(".fa","",genome))
  semi_join(., genome_counts_filt, by = "genome") %>% 
  arrange(match(genome,genome_counts_filt$genome))%>% 
  mutate(classification = str_replace_all(classification, ".__", "")) %>%
  separate(col = classification, sep = ";", into = c("domain", "phylum", "class", "order", "family", "genus", "species"))%>% 
  left_join(., genomes_length, by=join_by(genome==genome))%>%
  mutate(phylum = case_when(
        phylum == "Actinobacteriota" ~ "Actinomycetota",
        phylum == "Firmicutes" ~ "Bacillota",
        phylum == "Firmicutes_A" ~ "Bacillota_A",
        phylum == "Firmicutes_C" ~ "Bacillota_C",
        phylum == "Cyanobacteria" ~ "Cyanobacteriota",
        phylum == "Proteobacteria" ~ "Pseudomonadota",
        TRUE ~ phylum))
```

## Genome tree
```{r loadtree, comment="", message=FALSE, warning=FALSE, eval=FALSE}
genome_tree <- read_tree("resources/metagenomics/coassembly_species_gtdbtk_full.tree")
genome_tree$tip.label <- gsub(".fa","",genome_tree$tip.label) #remove .fa extension
genome_tree <- keep.tip(genome_tree, tip=genome_metadata$genome) # keep only MAG tips
```

## Preprocessing info
```{r prepro, comment="", message=FALSE, warning=FALSE, eval=FALSE}
unmapped <- read_tsv("resources/metagenomics/all_species_mapping_rate.txt")
mapping <- unmapped[c(1),] %>% 
 t() %>%
  row_to_names(row_number = 1) %>% 
 as.data.frame() %>%
  mutate_if(is.character, as.numeric) %>% 
  rownames_to_column("sample") %>% 
  mutate(mapped=100-unmapped)

preprocess_info <- read_tsv("resources/metagenomics/preprocessing_report.tsv") %>% 
  arrange(match(sample,sample_metadata$sample)) %>% 
  mutate(host_percent=100*(host_reads)/(reads_post_filt)) %>% 
  left_join(., mapping, by=join_by(sample==sample))

preprocess_info <- preprocess_info[match(sample_metadata$sample, preprocess_info$sample), ]
```

## Genome annotations
```{r annotation, comment="", message=FALSE, warning=FALSE, eval=FALSE}
genome_annotations <- read_tsv("resources/metagenomics/all_annotations.tsv.xz")
```

## Genome functions distillation
```{r distillation, comment="", message=FALSE, warning=FALSE, eval=FALSE}
genome_gifts <- distill(genome_annotations,GIFT_db,genomecol=2,annotcol=c(9,10,19), verbosity=F)
genome_gifts <- genome_gifts[rownames(genome_gifts) %in% genome_counts_filt$genome,]
genome_counts_filt$genome %in% rownames(genome_gifts) 
rownames(genome_gifts) %in% genome_counts_filt$genome
```

## Phyloseq object
```{r nozero_phylo, comment="", message=FALSE, warning=FALSE, eval=FALSE}
#phyloseq object without structural zeros
phylo_samples <- sample_metadata %>% 
  column_to_rownames("sample") %>% 
  sample_data() #convert to phyloseq sample_data object

phylo_counts <- genome_counts_filt %>% 
  column_to_rownames("genome") %>% 
  otu_table(., taxa_are_rows = TRUE) #convert to phyloseq otu_table object

phylo_taxonomy <- genome_metadata %>% 
  dplyr::select(domain,phylum,class,order,family,genus,species,genome,length) %>%
  column_to_rownames("genome") %>% 
  as.matrix() %>% 
  tax_table() 

tree <- phy_tree(genome_tree)

genome_data <- phyloseq(phylo_counts, phylo_taxonomy, phylo_samples, tree)
```

## Prepare color scheme
```{r get_ehi_colors_skin, warning=FALSE, comments="", message=FALSE, eval=FALSE}
phylum_colors <- read_tsv("https://raw.githubusercontent.com/earthhologenome/EHI_taxonomy_colour/main/ehi_phylum_colors.tsv") %>%
  mutate_at(vars(phylum), ~ str_replace(., "[dpcofgs]__", ""))%>%
    right_join(genome_metadata, by=join_by(phylum == phylum)) %>%
    arrange(match(genome, genome_tree$tip.label)) %>%
    select(phylum, colors) %>% 
    unique() %>%
    arrange(phylum) %>%
    pull(colors, name=phylum)
```

```{r savedata, comment="", message=FALSE, warning=FALSE, eval=FALSE}
save(sample_metadata, 
     genome_metadata, 
     read_counts, 
     genome_counts,
     genome_counts_filt,
     preprocess_info,
     genome_tree, 
     genome_gifts,
     genome_data,
     phylum_colors,
     file = "resources/metagenomics/data.Rdata")
```

# Amplicon
## Load data
```{r ampli, comment="", message=FALSE, warning=FALSE, eval=FALSE}
read_counts <- read_tsv("resources/amplicon/ASVs_count_overlap10_nochim0.tsv") %>% dplyr::rename(asv = 1) %>% 
  as.data.frame()
#asvs <- read_counts$asv
taxonomy_sp <- read_tsv("resources/amplicon/taxonomy_species_tax_overlap10_nochim0.tsv") %>%
  dplyr::rename(asv = 1) %>%
  as.data.frame()
hierarchy <- read_csv("resources/amplicon/metadata.csv")
samples <- read_counts %>% select(-asv) %>% colnames()
hierarchy <- hierarchy[hierarchy$Sample %in% samples,]
metadata <- tibble::column_to_rownames(hierarchy, "Sample")
read_counts_row <- column_to_rownames(read_counts, "asv")
taxonomy_table_row <- column_to_rownames(taxonomy_sp, "asv")
```

# Create a phyloseq object
```{r amplidata, comment="", message=FALSE, warning=FALSE, eval=FALSE}
#Phyloseq object
count_phy <- otu_table(read_counts_row, taxa_are_rows=T)
sample_info_tab_phy <- sample_data(metadata)
asv_tax.m <- as.matrix(taxonomy_table_row)
TAX = tax_table(asv_tax.m)
physeq.raw = phyloseq(count_phy, TAX, sample_info_tab_phy)
physeq.raw <- prune_taxa(taxa_sums(physeq.raw)>0, physeq.raw)
physeq.raw <- prune_samples(sample_sums(physeq.raw) > 0, physeq.raw)

physeq_sample <- subset_samples(physeq.raw, Control == "True sample")
physeq_sample <- prune_taxa(taxa_sums(physeq_sample)>0, physeq_sample)
```

# Remove contaminants
```{r decotam, comment="", message=FALSE, warning=FALSE, eval=FALSE}
sample_data(physeq.raw)$is.neg <- sample_data(physeq.raw)$Control == "Control sample"
contamdf.prev <- isContaminant(physeq.raw, method="prevalence", neg="is.neg")
table(contamdf.prev$contaminant)
contam_asvs <- row.names(contamdf.prev[contamdf.prev$contaminant == TRUE, ])
contaminants <- asv_tax.m[row.names(asv_tax.m) %in% contam_asvs, ]
```

```{r phylodec, comment="", message=FALSE, warning=FALSE, eval=FALSE}
# making new phyloseq object without ASV contaminats
physeq <- subset_samples(physeq.raw, Control == "True sample")
physeq <- prune_taxa(taxa_sums(physeq)>0, physeq)
goodTaxa <- setdiff(taxa_names(physeq), contam_asvs)
physeq.final <- prune_taxa(goodTaxa, physeq)
```

## Basic filtering
```{r basicfilt, comment="", message=FALSE, warning=FALSE, eval=FALSE}
read_counts_final <- data.frame(physeq.final@otu_table) 
count_filtdepth <- read_counts_final[,colSums(read_counts_final)>1000]
count_phy_filt <- otu_table(count_filtdepth, taxa_are_rows=T)

otu_table(physeq.final) <- count_phy_filt

#TAXONOMY filtering
physeq_bacteria <- subset_taxa(physeq.final, Kingdom == "Bacteria")
physeq_others <- subset_taxa(physeq.final, Kingdom != "Bacteria")
physeq_archaea <- subset_taxa(physeq.final, Kingdom == "Archaea")
physeq_phylum <- subset_taxa(physeq_bacteria, !is.na(Phylum))
physeq_phylumNA <- subset_taxa(physeq_bacteria, is.na(Phylum))
physeq_class <- subset_taxa(physeq_phylum, !is.na(Class))
physeq_classNA <- subset_taxa(physeq_phylum, is.na(Class))
```

## Removing chloroplast and mitochondria
```{r chlormito, comment="", message=FALSE, warning=FALSE, eval=FALSE}
taxonomy <- physeq.final@tax_table
is.mitochondria <- physeq.final@tax_table[,"Family"] %in% "Mitochondria"
taxonomy <- taxonomy[!is.mitochondria,]
is.chloroplast <- taxonomy[,"Order"] %in% "Chloroplast"
taxonomy <- taxonomy[!is.chloroplast,]
taxonomy.m <- as.matrix(taxonomy)
tax_table(physeq.final) <- taxonomy.m
physeq.final.filt <- prune_taxa(taxa_sums(physeq.final)>0, physeq.final)
```

## Save
```{r save16S, comment="", message=FALSE, warning=FALSE, eval=FALSE}
saveRDS(physeq.final.filt,"resources/amplicon/physeq_final_nocopyfilter.RData")
```

# Merging amplicon and shotgun data

```{r load16Sdata, comment="", message=FALSE, warning=FALSE, eval=FALSE}
physeq_16S <- readRDS("resources/amplicon/physeq_final_nocopyfilter.RData")
physeq_16S <- prune_taxa(taxa_sums(physeq_16S)>0, physeq_16S)
```

```{r cleandata, comment="", message=FALSE, warning=FALSE, eval=FALSE}
#list duplicated data
remov_samples <- c("P19_r1", "P19_r2", "P10_r2", "P10_r1", "E06_r3", "E06_r2", "E15_r1", "E29_r1", "E29_r2", "E37_r2", "E37_r3", "E47_r2", "E50_r1", "H09_r1", "H13_r2", "H39_r1", "P35_r3", "P64_r2", "P64_r3")

#Change the names in asv_table
asv_16S <- data.frame(physeq_16S@otu_table) 
names(asv_16S) <- sub("^[^.]+\\.", "",names(asv_16S))
colnames(asv_16S)
#filter duplicated data
asv_16S_filt <- asv_16S[, -which(names(asv_16S) %in% remov_samples)]
#remove _r* from the names
names(asv_16S_filt) <- gsub(pattern = "_r[1-3]$", "", x = names(asv_16S_filt))
```
```{r filterdata, comment="", message=FALSE, warning=FALSE, eval=FALSE}
asv_16S_filt <- asv_16S_filt[rowSums(asv_16S_filt[, -1] > 0) != 0, ] %>% 
  rownames_to_column(., "asv")
```

```{r sampledata, comment="", message=FALSE, warning=FALSE, eval=FALSE}
#Prepare the samples data and filter
sample_metadata_16S <-  data.frame(physeq_16S@sam_data) %>% 
  rownames_to_column(., "sample_id") %>% 
  separate_wider_delim(sample_id, ".", names = c('id', 'sample')) %>% 
  select(-id) %>% 
  separate_wider_delim(Sp_Biome, "_", names = c('Sp', 'Biome')) %>%  
  column_to_rownames(., "sample") 

# Define the row indices to keep based on row names
rows_to_remove <- !(rownames(sample_metadata_16S) %in% remov_samples)
# Filter the data frame based on the row indices
sample_metadata_16S <- sample_metadata_16S[rows_to_remove, ]

#remove _r*
rownames(sample_metadata_16S) <- gsub(pattern = "_r[1-3]$", "", x = rownames(sample_metadata_16S))

sample_metadata_16S <- sample_metadata_16S %>% 
  rownames_to_column(., "sample")

genome_metadata_16S <-  data.frame(physeq_16S@tax_table) %>% 
  rownames_to_column(., "asv") %>% 
  filter(asv %in% asv_16S_filt$asv)
```

```{r}
save(sample_metadata_16S, 
     asv_16S_filt, 
     genome_metadata_16S,
     file = "resources/amplicon/data_withoutcopy.Rdata")
```

```{r comparabledata, comment="", message=FALSE, warning=FALSE}
#keep samples present in metagenomic study
sample_metadata_16S_filt <- sample_metadata_16S[sample_metadata_16S$sample %in% sample_metadata$sample,]

#keep samples present in 16S study
sample_metagenomic_filt <- sample_metadata[sample_metadata$sample %in% sample_metadata_16S$sample,]
```

## Prepare the metagenomic data
```{r metagenomicdata, comment="", message=FALSE, warning=FALSE, eval=FALSE}
phyloM_samples <- sample_metagenomic_filt %>%
  column_to_rownames(., "sample") %>% 
  sample_data() #convert to phyloseq sample_data object

sam_data(genome_data) <- phyloM_samples
phyloseq_metagenome_filtered <- prune_taxa(taxa_sums(genome_data)>0, genome_data)

sample_metadata <- data.frame(phyloseq_metagenome_filtered@sam_data)%>% 
  rownames_to_column(., "sample")
genome_metadata <- data.frame(phyloseq_metagenome_filtered@tax_table) %>% 
  rownames_to_column(., "genome")
genome_counts_filt <- data.frame(phyloseq_metagenome_filtered@otu_table) %>% 
  rownames_to_column(., "genome")

preprocess_info <- preprocess_info %>% 
  filter(sample %in% sample_metadata$sample)

genome_tree <- keep.tip(genome_tree, tip=genome_metadata$genome) 

genome_gifts <- genome_gifts[rownames(genome_gifts) %in% genome_counts_filt$genome,]

save(sample_metadata, 
     genome_metadata, 
     read_counts, 
     genome_counts,
     genome_counts_filt,
     preprocess_info,
     genome_tree, 
     genome_gifts,
     phyloseq_metagenome_filtered,
     phylum_colors,
     file = "resources/metagenomics/data_filtered.Rdata")
```

## Prepare the 16S data
```{r 16Sdata, comment="", message=FALSE, warning=FALSE, eval=FALSE}

phylo_samples <- sample_metadata_16S_filt %>% 
  remove_rownames() %>% 
  column_to_rownames(., "sample") %>% 
  sample_data() 

phylo_counts <- asv_16S_filt %>% 
  column_to_rownames(., "asv") %>% 
  otu_table(., taxa_are_rows = TRUE) 

phylo_taxonomy <- genome_metadata_16S %>% 
  column_to_rownames(., "asv")%>% 
  as.matrix() %>% 
  tax_table() 

physeq_16S_filt <- phyloseq(phylo_counts,  phylo_samples, phylo_taxonomy)
phyloseq_16S_filtered <- prune_taxa(taxa_sums(physeq_16S_filt)>0, physeq_16S_filt)

sample_metadata <- data.frame(phyloseq_16S_filtered@sam_data)%>% 
  rownames_to_column(., "sample") %>% 
  rename(species_name=Species) %>% 
  rename(Species=Sp) %>%
  mutate(Habitat = ifelse(Habitat == "Natural", "Nat", Habitat)) %>%
  mutate(Habitat = ifelse(Habitat == "Anthropogenic", "Anth", Habitat))

genome_metadata <- data.frame(phyloseq_16S_filtered@tax_table) %>% 
  rownames_to_column(., "genome") %>% 
  mutate(Phylum = case_when(
        Phylum == "Actinobacteriota" ~ "Actinomycetota",
        Phylum == "Firmicutes" ~ "Bacillota",
        Phylum == "Firmicutes_A" ~ "Bacillota_A",
        Phylum == "Firmicutes_C" ~ "Bacillota_C",
        Phylum == "Cyanobacteria" ~ "Cyanobacteriota",
        Phylum == "Proteobacteria" ~ "Pseudomonadota",
        TRUE ~ Phylum))

colnames(genome_metadata)<- c("genome", "domain", "phylum", "class", "order", "family", "genus", "species")
genome_counts_filt <- data.frame(phyloseq_16S_filtered@otu_table) %>% 
  rownames_to_column(., "genome")

## Prepare color scheme
phylum_colors <- read_tsv("https://raw.githubusercontent.com/earthhologenome/EHI_taxonomy_colour/main/ehi_phylum_colors.tsv") %>%
  mutate_at(vars(phylum), ~ str_replace(., "[dpcofgs]__", ""))%>%
    right_join(genome_metadata, by=join_by(phylum == phylum)) %>%
    arrange(match(genome, genome_metadata$genome)) %>%
    select(phylum, colors) %>% 
    unique() %>%
    arrange(phylum) %>%
    pull(colors, name=phylum)

taxonomy <- genome_metadata %>% 
  column_to_rownames("genome") %>% 
  as.matrix() %>% 
  tax_table()
sample_data <- sample_metadata %>% 
  column_to_rownames("sample") %>% 
  sample_data()

tax_table(phyloseq_16S_filtered) <- taxonomy

sample_data(phyloseq_16S_filtered) <- sample_data

```
```{r save_ampli, comment="", message=FALSE, warning=FALSE, eval=FALSE}
save(sample_metadata, 
     genome_metadata,
     genome_counts_filt,
     phyloseq_16S_filtered,
     phylum_colors,
     file = "resources/amplicon/data_nocopyfilt.Rdata")
```

## Create master table for amplicon data

```{r load_data, comment="", message=FALSE, warning=FALSE, eval=FALSE}
load("resources/amplicon/data_nocopyfilt.Rdata")
```

```{r master_table, comment="", message=FALSE, warning=FALSE, eval=FALSE}
threshold_table <- genome_counts_filt %>% 
  mutate_at(vars(-genome), ~ ./sum(.)) %>%
  pivot_longer(!genome, names_to = "sample",values_to = "t0") %>%
  left_join(sample_metadata %>% select(sample,Species), by="sample") %>%
  mutate(t1 = if_else(t0 < 0.0001, 0, t0),
         t2 = if_else(t0 < 0.001, 0, t0),
         t3 = if_else(t0 < 0.01, 0, t0),
         t4 = if_else(t0 < 0.1, 0, t0)) %>%
  pivot_longer(!c(genome,sample,Species), names_to = "copy_threshold",values_to = "value") 

prevalence_table <- threshold_table %>%
  group_by(Species,genome,copy_threshold) %>% 
  summarise(present = sum(value != 0, na.rm = TRUE),
            total = n()) %>% 
  mutate(prevalence=present/total) %>% 
  select(-present,-total)

master_table <- inner_join(threshold_table,prevalence_table,by=join_by("Species"=="Species","genome"=="genome","copy_threshold"=="copy_threshold")) %>% 
  rename(p0=value) %>% 
  mutate(p10 = if_else(prevalence < 0.1, 0, p0),
         p20 = if_else(prevalence < 0.2, 0, p0),
         p30 = if_else(prevalence < 0.3, 0, p0),
         p40 = if_else(prevalence < 0.4, 0, p0)) %>%
  select(-prevalence) %>% 
  pivot_longer(!c(genome,sample,Species,copy_threshold), names_to = "prevalence",values_to = "value") %>% 
  mutate(pa = if_else(value > 0, 1, 0))
```

```{r save_master_table, comment="", message=FALSE, warning=FALSE, eval=FALSE}
save(master_table,
     file = "resources/amplicon/master_table.Rdata")
```